pathmux/viztree.go

```diff
@@ -0,0 +1,148 @@
+package pathmux
+
+// Exploded version of the pathmux tree designed for being easy to use in a visualization.
+// Simple wildcard nodes are represented by the ':' prefix and free wildcard nodes with the '*' prefix.
+type VizTree struct {
+	Path     string // string representation of the node path
+	children []*VizTree
+	canMatch bool
+
+	// The wildcard names still waiting to be processed, should be always nil after the end of the conversion.
+	leafWildcardNames []string
+}
+
+// Creates a new visualization tree from a pathmux.Tree.
+func NewVizTree(tree *Tree) *VizTree {
+	vizTree := aggregateTree((*node)(tree), "/")
+	return vizTree[0]
+}
+
+// Return a child referenced by direct Path
+//
+// Example:
+//
+// - in a tree such as: / -> images -> abc
+//
+// looking for images in the root tree would yield the images subtree.
+func (n *VizTree) Child(path string) *VizTree {
+	return findNode(n.children, path)
+}
+
+func aggregateTree(tree *node, previousPath string) []*VizTree {
+	if tree == nil {
+		return nil
+	}
+	var currentNode, middleNode *VizTree
+	nextPath := previousPath + tree.path
+	// On every slash we create new nodes if the previous Path has a value, otherwise we ignore it
+	if tree.path == "/" {
+		if previousPath != "" {
+			currentNode = &VizTree{previousPath, nil, tree.leafValue != nil, tree.leafWildcardNames}
```

Comment:
with this many fields, it is more readable to write out the field names in the struct literal. Instead of:

```
&VizTree{nextPath, nil, true, tree.leafWildcardNames}
```

it would better if:

```
&VizTree{
    Path: nextPath,
    canMatch: true,
    leafWildcardNames: tree.leafWildcardNames,
}
```